h2/src/main/org/h2/mvstore/tx/TransactionMap.java

```diff
@@ -99,76 +99,72 @@ public long sizeAsLongMax() {
+            mapRootReference = map.getRoot();
+            undoLogRootReference = store.undoLog.getRoot();
+        } while(committingTransactions != store.committingTransactions.get() ||
+                mapRootReference != map.getRoot());
+
+        Page undoRootPage = undoLogRootReference.root;
+        long undoLogSize = undoRootPage.getTotalCount();
+        Page mapRootPage = mapRootReference.root;
+        long size = mapRootPage.getTotalCount();
+        if (undoLogSize == 0) {
+            return size;
+        }
+        if (undoLogSize > size) {
+            // the undo log is larger than the map -
+            // count the entries of the map
+            size = 0;
+            Cursor<K, VersionedValue> cursor = map.cursor(null);
+            while (cursor.hasNext()) {
+                K key = cursor.next();
+                VersionedValue data = cursor.getValue();
+                data = getValue(mapRootPage, undoRootPage, key, readLogId, data, committingTransactions);
+                if (data != null && data.value != null) {
+                    size++;
                 }
-                return size;
             }
-            // the undo log is smaller than the map -
-            // scan the undo log and subtract invisible entries
-            synchronized (undo) {
-                // re-fetch in case any transaction was committed now
-                long size = map.sizeAsLong();
-                MVMap<Object, Integer> temp = transaction.store
-                        .createTempMap();
-                try {
-                    for (Map.Entry<Long, Object[]> e : undo.entrySet()) {
-                        Object[] op = e.getValue();
-                        int m = (Integer) op[0];
-                        if (m != mapId) {
-                            // a different map - ignore
-                            continue;
-                        }
-                        @SuppressWarnings("unchecked")
-                        K key = (K) op[1];
-                        if (get(key) == null) {
-                            Integer old = temp.put(key, 1);
-                            // count each key only once (there might be
-                            // multiple
-                            // changes for the same key)
-                            if (old == null) {
-                                size--;
-                            }
-                        }
+            return size;
+        }
+        // the undo log is smaller than the map -
+        // scan the undo log and subtract invisible entries
+        MVMap<Object, Integer> temp = transaction.store
+                .createTempMap();
+        try {
+            Cursor cursor = new Cursor<Long, Object[]>(undoRootPage, null);
+            while (cursor.hasNext()) {
+                cursor.next();
+                Object[] op = (Object[]) cursor.getValue();
+                int m = (Integer) op[0];
```

Comment:
With proper declaration of `Cursor` cast to `Object[]` is not needed. Cast to `int` also looks more natural than current cast to `Integer`, but it does not matter.